pkg/arvo/app/chat-view.hoon

```diff
@@ -0,0 +1,328 @@
+    :_  this
+    :~  [ost.bol %peer / [our.bol %chat-store] /updates]
+        [ost.bol %connect / [~ /'~chat'] %chat-view]
+        (launch-poke [/configs '/~chat/js/tile.js'])
+    ==
+  [~ this]
+::
+++  bound
+  |=  [wir=wire success=? binding=binding:eyre]
+  ^-  (quip move _this)
+  [~ this]
+::
+++  poke-handle-http-request
+  %-  (require-authorization:app ost.bol move this)
+  |=  =inbound-request:eyre
+  ^-  (quip move _this)
+  ::
+  =+  url=(parse-request-line url.request.inbound-request)
+  =/  name=@t
+    =+  back-path=(flop site.url)
+    ?~  back-path
+      ''
+    i.back-path
+  ?:  =(name 'tile')
+    [[ost.bol %http-response (js-response:app tile-js)]~ this]
+  ?+  site.url
+    :_  this
+    [ost.bol %http-response not-found:app]~
+  ::
+  ::  styling
+  ::
+      [%'~chat' %css %index ~]
+    :_  this
+    [ost.bol %http-response (css-response:app style)]~
+  ::
+  ::  javascript
+  ::
+      [%'~chat' %js %index ~]
+    :_  this
+    [ost.bol %http-response (js-response:app script)]~
+  ::
+  ::  images
+  ::
+      [%'~chat' %img *]
+    =/  img  (as-octs:mimes:html (~(got by chat-png) `@ta`name))
+    :_  this
+    [ost.bol %http-response (png-response:app img)]~
+  ::
+    [%'~chat' %paginate @t @t *]
+    =/  start  (need (rush i.t.t.site.url dem))
+    =/  end  (need (rush i.t.t.t.site.url dem))
+    =/  pax  t.t.t.t.site.url
+    =/  envelopes  (envelope-scry [(scot %ud start) (scot %ud end) pax])
+    :_  this
+    :~
+      :+  ost.bol
+        %http-response
+      %-  json-response:app
+      %-  json-to-octs 
+      %+  envelopes-update
+        envelopes
+      [start end pax]
+    ==
+  ::
+  ::  inbox page
+  ::
+     [%'~chat' *]
+    :_  this
+    [ost.bol %http-response (html-response:app index)]~
+  ==
+::
+++  poke-json
+  |=  jon=json
+  ^-  (quip move _this)
+  ?.  =(src.bol our.bol)
+    [~ this]
+  (poke-chat-view-action (json-to-view-action jon))
+::
+++  poke-chat-view-action
+  |=  act=chat-view-action
+  ^-  (quip move _this)
+  ?.  =(src.bol our.bol)
+    [~ this]
+  ?-  -.act
+      %create
+    ::  TODO: add invites
+    =/  pax  [(scot %p our.bol) path.act]
+    =/  group-read=path  [%chat (weld pax /read)]
+    =/  group-write=path  [%chat (weld pax /write)]
+    :_  this
+    %+  weld
+      :~  (chat-poke [%create our.bol path.act])
+          (group-poke [%bundle group-read])
+          (group-poke [%bundle group-write])
+          (group-poke [%add read.act group-read])
+          (group-poke [%add write.act group-write])
+          (chat-hook-poke [%add-owned pax security.act])
+      ==
+    (create-security [%chat pax] security.act)
+  ::
+      %delete
+    =/  group-read  [%chat (weld path.act /read)]
+    =/  group-write  [%chat (weld path.act /write)]
+    :_  this
+    :~  (chat-hook-poke [%remove path.act])
+        (group-poke [%unbundle group-read])
+        (group-poke [%unbundle group-write])
+        (chat-poke [%delete path.act])
+    ==
+  ::
+  ==
+::
+++  peer-initial
```

Comment:
What's here should work most of the time, but FWIW the interaction between `/initial` and `/updates` isn't quite perfect (at least as used in `subscription.js`).  If a message comes in after the subscription to `/initial` is processed but before the subscription to `/updates` is, then it'll be skipped altogether.

We've done "get limited backlog and future updates" poorly in the past.  The two guarantees you'd like to have are: (1) updates come on a single canonical path and (2) never miss an update after the first message in the backlog.  The old `/updates/<from-number>` approach violates (1), and this approach can occasionally violate (2).  I believe the platonic form is basically:

- subscribe to `/updates`.  This immediately gives you a message number `n`, representing the number of messages in the channel.  It also gives you any new updates as they come in.
- subscribe to `/initial/(scot %ud (sub n 100))/(scot %ud n)`, which responds immediately with messages between those two numbers.

This way, you're already listening to `/updates` by the time you're getting the backlog, so you'll never miss a message.  The disadvantage of this approach is that you have to do the subscriptions consecutively and potentially reorder messages if `/updates` gives you results before `/initial`, so the client code is a bit more complicated.

Reply:
Seems like I'd much prefer to just not cut off the subscription to %chat-view, and continue ferrying updates along through a single %chat-view subscription, since %chat-view is already subscribed to the /updates path.

Reply:
This approach was successful and cleaner. Thanks for the motivation to change it!